File: apps/linux/src/bluetooth.rs
//! Bluetooth transport: framed E2E envelopes over RFCOMM / BLE GATT.
//!
//! Mirrors the framing in `packages/protocol/src/lan.ts`
//! (`UCM1 <seq>/<total> <base64>` lines, `BT_MTU_CHUNK = 512`).
//!
//! Architecture: this module owns the *framing + envelope validation* and a
//! stream-based transfer (`write_envelope` / `read_envelope`) that works over
//! ANY reliable byte stream — RFCOMM socket, BLE L2CAP CoC, or a TCP socket
//! in tests. The BlueZ adapter setup itself (advertise/scan/pair) is
//! intentionally behind [`BtAdapter`], whose default implementation probes
//! the real stack (`bluetoothctl`, `/sys/class/bluetooth`) and reports a
//! structured [`BtStatus`] instead of failing the daemon when no radio
//! exists (CI, VMs, containers). Swapping in the `bluer` crate later only
//! means implementing `BtAdapter` — framing stays untouched.

use anyhow::{Context, Result};
use tokio::io::{AsyncBufReadExt, AsyncRead, AsyncWrite, AsyncWriteExt, BufReader};
use tracing::{info, warn};

use crate::lan::{validate_envelope, LanEnvelope};

/// Must match `BT_SERVICE_UUID` in `@ucm/protocol` + Android `lib/bluetooth.ts`.
pub const SERVICE_UUID: &str = "7c9e5f2a-9b3d-4a5e-9f2c-1a2b3c4d5e6f";
/// Must match `BT_CHAR_UUID` in `@ucm/protocol`.
pub const CHAR_UUID: &str = "7c9e5f2b-9b3d-4a5e-9f2c-1a2b3c4d5e6f";
/// Must match `BT_MTU_CHUNK` (conservative for BLE writes).
pub const MTU_CHUNK: usize = 512;
pub const FRAME_PREFIX: &str = "UCM1";

#[derive(Debug, Clone)]
pub struct BtStatus {
    pub available: bool,
    pub powered: bool,
    pub detail: String,
}

/// Probe the host Bluetooth stack without taking exclusive ownership.
/// Returns `available=false` (with a human-readable `detail`) on headless
/// CI/VMs instead of erroring, so `ucm daemon` keeps running on WiFi/cloud.
pub fn availability() -> BtStatus {
    if !std::path::Path::new("/sys/class/bluetooth").exists() {
        // `bluetoothctl` may still exist via USB dongles that haven't
        // registered sysfs yet — check both before giving up.
        match std::process::Command::new("bluetoothctl").arg("--version").output() {
            Ok(o) if o.status.success() => {
                let v = String::from_utf8_lossy(&o.stdout).trim().to_string();
                return BtStatus { available: true, powered: false, detail: format!("radio sysfs missing, {v} present — dongle may be unplugged") };
            }
            _ => {
                return BtStatus {
                    available: false,
                    powered: false,
                    detail: "no /sys/class/bluetooth and no bluetoothctl — Bluetooth disabled or unavailable (WiFi/cloud still work)".into(),
                }
            }
        }
    }
    let powered = std::process::Command::new("bluetoothctl")
        .args(["show"])
        .output()
        .map(|o| String::from_utf8_lossy(&o.stdout).contains("Powered: yes"))
        .unwrap_or(false);
    BtStatus {
        available: true,
        powered,
        detail: if powered {
            format!("adapter ready; advertise service {SERVICE_UUID} via BlueZ (see `ucm bt-status`)")
        } else {
            "adapter present but not powered — run `bluetoothctl power on`".into()
        },
    }
}

/// Split envelope JSON into MTU-sized `UCM1 seq/total base64` frame lines.
pub fn encode_frames(env: &LanEnvelope) -> Result<Vec<String>> {
    let raw = serde_json::to_vec(env)?;
    let b64 = base64::Engine::encode(&base64::engine::general_purpose::STANDARD, raw);
    let step = (MTU_CHUNK / 4 * 4).max(4);
    let mut parts = Vec::new();
    let mut i = 0;
    while i < b64.len() {
        let end = (i + step).min(b64.len());
        parts.push(b64[i..end].to_string());
        i = end;
    }
    if parts.is_empty() {
        parts.push(String::new());
    }
    let total = parts.len();
    Ok(parts
        .into_iter()
        .enumerate()
        .map(|(k, p)| format!("{FRAME_PREFIX} {}/{total} {p}", k + 1))
        .collect())
}

/// Reassemble + validate frames (order-independent, duplicates rejected).
pub fn decode_frames(lines: &[String]) -> Result<LanEnvelope> {
    if lines.is_empty() {
        anyhow::bail!("no frames");
    }
    let mut chunks: Vec<Option<String>> = Vec::new();
    let mut total: Option<usize> = None;
    for line in lines {
        let rest = line.strip_prefix(&format!("{FRAME_PREFIX} ")).context("bad frame prefix")?;
        let (seq_total, payload) = rest.split_once(' ').context("bad frame shape")?;
        let (seq_s, total_s) = seq_total.split_once('/').context("bad frame seq")?;
        let seq: usize = seq_s.parse().context("bad seq")?;
        let t: usize = total_s.parse().context("bad total")?;
        match total {
            None => {
                total = Some(t);
                chunks.resize(t, None);
            }
            Some(prev) if prev != t => anyhow::bail!("mixed frame totals"),
            _ => {}
        }
        if seq < 1 || seq > t {
            anyhow::bail!("frame seq out of range");
        }
        if chunks[seq - 1].is_some() {
            anyhow::bail!("duplicate frame");
        }
        chunks[seq - 1] = Some(payload.to_string());
    }
    let total = total.unwrap_or(0);
    if total == 0 {
        anyhow::bail!("no frames");
    }
    if chunks.iter().any(|c| c.is_none()) {
        anyhow::bail!("incomplete frames");
    }
    let joined: String = chunks.into_iter().flatten().collect();
    use base64::{engine::general_purpose::STANDARD as B64, Engine};
    let raw = B64.decode(&joined).context("frame base64")?;
    let env: LanEnvelope = serde_json::from_slice(&raw).context("envelope json")?;
    validate_envelope(&env)?;
    if env.transport != "bluetooth" {
        anyhow::bail!("expected bluetooth envelope");
    }
    Ok(env)
}

/// Write one envelope as newline-delimited frames on any byte stream
/// (RFCOMM socket, L2CAP CoC, test TCP). Ends with a blank line sentinel.
pub async fn write_envelope<W>(writer: &mut W, env: &LanEnvelope) -> Result<usize>
where
    W: AsyncWrite + Unpin,
{
    let frames = encode_frames(env)?;
    let n = frames.len();
    for f in &frames {
        writer.write_all(f.as_bytes()).await?;
        writer.write_all(b"\n").await?;
    }
    writer.write_all(b"\n").await?; // end-of-envelope sentinel
    writer.flush().await?;
    Ok(n)
}

/// Read frames until the blank-line sentinel, then reassemble + validate.
/// `max_frames` bounds memory against a hostile peer (default 512 ≈ 256 KiB).
pub async fn read_envelope<R>(reader: &mut BufReader<R>, max_frames: usize) -> Result<LanEnvelope>
where
    R: AsyncRead + Unpin,
{
    let mut lines: Vec<String> = Vec::new();
    loop {
        let mut line = String::new();
        let n = reader.read_line(&mut line).await.context("bt read")?;
        if n == 0 {
            if lines.is_empty() {
                anyhow::bail!("peer closed connection");
            }
            break;
        }
        let trimmed = line.trim().to_string();
        if trimmed.is_empty() {
            break; // sentinel
        }
        lines.push(trimmed);
        if lines.len() > max_frames {
            anyhow::bail!("envelope too many frames");
        }
    }
    decode_frames(&lines)
}

/// Send over an established stream and log at info level (no plaintext).
pub async fn send_over_stream<S>(stream: &mut S, env: &LanEnvelope) -> Result<usize>
where
    S: AsyncWrite + Unpin,
{
    let n = write_envelope(stream, env).await?;
    info!(frames = n, id = %env.item.id, "bt envelope sent");
    Ok(n)
}

/// Adapter hook for real BlueZ work (advertise/scan/connect). The default
/// `NoopAdapter` keeps the daemon healthy where no radio exists; a future
/// `bluer`-backed impl slots in here without touching framing or sync.
#[async_trait::async_trait]
pub trait BtAdapter: Send + Sync {
    fn status(&self) -> BtStatus;
    async fn advertise(&self) -> Result<()>;
    async fn scan(&self, secs: u64) -> Result<Vec<DiscoveredBtPeer>>;
}

#[derive(Debug, Clone)]
pub struct DiscoveredBtPeer {
    pub address: String,
    pub name: Option<String>,
    pub service_found: bool,
}

pub struct NoopAdapter;

#[async_trait::async_trait]
impl BtAdapter for NoopAdapter {
    fn status(&self) -> BtStatus {
        availability()
    }
    async fn advertise(&self) -> Result<()> {
        let st = availability();
        if !st.available {
            warn!(detail = %st.detail, "bt advertise skipped");
            return Ok(());
        }
        info!("bt advertise: register GATT service {SERVICE_UUID} via bluetoothctl/bluer (see docs/wifi-bluetooth.md)");
        Ok(())
    }
    async fn scan(&self, _secs: u64) -> Result<Vec<DiscoveredBtPeer>> {
        let st = availability();
        if !st.available {
            warn!(detail = %st.detail, "bt scan skipped");
            return Ok(vec![]);
        }
        info!("bt scan: use `bluetoothctl scan on` then `ucm bt-send --addr <MAC>` (native GATT client lands with the bluer backend)");
        Ok(vec![])
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::lan::LanItem;

    fn env() -> LanEnvelope {
        LanEnvelope {
            v: 1,
            transport: "bluetooth".into(),
            sender_device_id: "d1".into(),
            sender_name: Some("ubuntu".into()),
            item: LanItem {
                id: "123e4567-e89b-12d3-a456-426614174000".into(),
                owner_id: "u1".into(),
                source_device_id: "d1".into(),
                content_type: "text/plain".into(),
                ciphertext: base64::Engine::encode(&base64::engine::general_purpose::STANDARD, b"hello-bt"),
                nonce: base64::Engine::encode(&base64::engine::general_purpose::STANDARD, b"123456789012"),
                metadata: serde_json::json!({}),
                created_at: "2024-01-01T00:00:00.000Z".into(),
                expires_at: None,
                deleted_at: None,
            },
        }
    }

    #[test]
    fn frames_roundtrip_small_and_large() {
        let e = env();
        let f = encode_frames(&e).unwrap();
        assert_eq!(decode_frames(&f).unwrap().item.id, e.item.id);
        let mut big = e.clone();
        big.item.ciphertext = base64::Engine::encode(&base64::engine::general_purpose::STANDARD, vec![b'x'; 5000]);
        let frames = encode_frames(&big).unwrap();
        assert!(frames.len() > 1);
        let mut rev = frames.clone();
        rev.reverse();
        assert_eq!(decode_frames(&rev).unwrap().item.id, big.item.id);
    }

    #[test]
    fn frames_reject_corrupt() {
        let e = env();
        let f = encode_frames(&e).unwrap();
        assert!(decode_frames(&f[..f.len().saturating_sub(1)]).is_err() || f.len() == 1);
        assert!(decode_frames(&["UCM1 1/1 !!!".to_string()]).is_err());
        // wrong transport tag must fail even if framing is fine
        let mut wifi = e.clone();
        wifi.transport = "wifi".into();
        let fw = encode_frames(&wifi).unwrap();
        assert!(decode_frames(&fw).is_err());
    }

    #[tokio::test]
    async fn stream_roundtrip_over_tcp() {
        let e = env();
        let listener = tokio::net::TcpListener::bind("127.0.0.1:0").await.unwrap();
        let addr = listener.local_addr().unwrap();
        let srv = tokio::spawn(async move {
            let (sock, _) = listener.accept().await.unwrap();
            let mut r = BufReader::new(sock);
            read_envelope(&mut r, 512).await.unwrap()
        });
        let mut cli = tokio::net::TcpStream::connect(addr).await.unwrap();
        send_over_stream(&mut cli, &e).await.unwrap();
        let got = srv.await.unwrap();
        assert_eq!(got.item.id, e.item.id);
    }

    #[test]
    fn availability_never_panics() {
        let st = availability();
        assert!(!st.detail.is_empty());
    }
}

File: apps/android/lib/bluetooth.ts
/**
 * Bluetooth transport: framed E2E envelopes shared with the Linux agent.
 *
 * Frame format MUST match `packages/protocol/src/lan.ts` and
 * `apps/linux/src/bluetooth.rs`:
 *   `UCM1 <seq>/<total> <base64>` lines, 512B chunks, blank-line terminator.
 * The envelope JSON inside is the same ciphertext the WiFi LAN path
 * carries — framing adds no crypto, AES-GCM AAD still authenticates.
 *
 * Runtime: uses `react-native-ble-plx` when the dev build includes it
 * (optional peer dep — never imported statically so Expo Go / typecheck
 * keep working). Without it, `getBtStatus()` reports `unavailable` with
 * an install hint and `sendEnvelope` queues into an in-memory outbox
 * instead of throwing away the user's intent.
 */

export const BT_SERVICE_UUID = "7c9e5f2a-9b3d-4a5e-9f2c-1a2b3c4d5e6f";
export const BT_CHAR_UUID = "7c9e5f2b-9b3d-4a5e-9f2c-1a2b3c4d5e6f";
export const BT_MTU_CHUNK = 512;
export const BT_FRAME_PREFIX = "UCM1";

export interface BtEnvelope {
  v: 1;
  transport: "bluetooth";
  sender_device_id: string;
  sender_name?: string;
  // Opaque to transport; validated for shape only.
  item: Record<string, unknown>;
}

export type BtAvailability =
  | { available: true; backend: "ble-plx"; detail: string }
  | { available: false; backend: "none"; detail: string };

export function encodeBtFrames(envelope: BtEnvelope, chunkBytes = BT_MTU_CHUNK): string[] {
  const raw: string = JSON.stringify(envelope);
  // base64 the UTF-8 bytes (btoa path works in RN/Hermes without Buffer).
  const bytes = encodeUtf8(raw);
  const b64 = base64Of(bytes);
  const step = Math.max(4, Math.floor(chunkBytes / 4) * 4);
  const parts: string[] = [];
  for (let i = 0; i < b64.length; i += step) parts.push(b64.slice(i, i + step));
  const total = Math.max(1, parts.length);
  return parts.map((p, i) => `${BT_FRAME_PREFIX} ${i + 1}/${total} ${p}`);
}

export function decodeBtFrames(lines: string[]): BtEnvelope {
  if (lines.length === 0) throw new Error("no frames");
  const chunks: (string | undefined)[] = [];
  let total = -1;
  for (const rawLine of lines) {
    const line = rawLine.trim();
    const m = new RegExp(`^${BT_FRAME_PREFIX} (\\d+)/(\\d+) ([A-Za-z0-9+/=]*)$`).exec(line);
    if (!m) throw new Error(`bad frame: ${line.slice(0, 32)}`);
    const seq = parseInt(m[1], 10);
    const t = parseInt(m[2], 10);
    if (total === -1) {
      total = t;
      chunks.length = t;
    }
    if (t !== total) throw new Error("mixed frame totals");
    if (seq < 1 || seq > total) throw new Error("frame seq out of range");
    if (chunks[seq - 1] !== undefined) throw new Error("duplicate frame");
    chunks[seq - 1] = m[3];
  }
  if (total !== lines.length) throw new Error(`incomplete frames: ${lines.length}/${total}`);
  const joined = (chunks as string[]).join("");
  const env = JSON.parse(utf8Of(base64ToBytes(joined))) as BtEnvelope;
  if (env?.v !== 1 || env?.transport !== "bluetooth" || typeof env?.sender_device_id !== "string") {
    throw new Error("reassembled envelope invalid");
  }
  return env;
}

/** Probe for a BLE backend without a static import (Expo Go safe). */
export async function getBtStatus(): Promise<BtAvailability> {
  try {
    // Optional peer dep for dev builds; absent in Expo Go / web.
    const mod = "react-native-ble-plx";
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const req = (globalThis as unknown as { require?: (m: string) => unknown }).require;
    if (req) {
      req(mod);
      return {
        available: true,
        backend: "ble-plx",
        detail: `BLE ready; UCM service ${BT_SERVICE_UUID}`,
      };
    }
    await import(/* @vite-ignore */ mod);
    return {
      available: true,
      backend: "ble-plx",
      detail: `BLE ready; UCM service ${BT_SERVICE_UUID}`,
    };
  } catch {
    return {
      available: false,
      backend: "none",
      detail:
        "Bluetooth needs a dev build with `react-native-ble-plx` (Expo Go has no BLE). " +
        "WiFi LAN still works; see docs/wifi-bluetooth.md.",
    };
  }
}

// Outbox keeps user intent when no radio is present (flushed on reconnect).
const outbox: { deviceId: string; frames: string[]; queuedAt: string }[] = [];

export function btOutbox(): { deviceId: string; frames: string[]; queuedAt: string }[] {
  return [...outbox];
}

export async function sendEnvelopeViaBt(
  deviceId: string,
  envelope: BtEnvelope,
): Promise<{ frames: number; queued: boolean }> {
  const frames = encodeBtFrames(envelope);
  const st = await getBtStatus();
  if (!st.available) {
    outbox.push({ deviceId, frames, queuedAt: new Date().toISOString() });
    return { frames: frames.length, queued: true };
  }
  // Real GATT write path lives here once the dev-build BLE manager is
  // wired (connect deviceId → discover service BT_SERVICE_UUID → write
  // BT_CHAR_UUID per frame). Until then, queue so nothing is lost.
  outbox.push({ deviceId, frames, queuedAt: new Date().toISOString() });
  return { frames: frames.length, queued: true };
}

// --- minimal base64/utf8 without Node Buffer (RN/Hermes safe) ---

function encodeUtf8(s: string): Uint8Array {
  return TextEncoder
    ? new TextEncoder().encode(s)
    : Uint8Array.from([...s].map((c) => c.charCodeAt(0) & 0xff));
}

function utf8Of(bytes: Uint8Array): string {
  if (typeof TextDecoder !== "undefined") return new TextDecoder().decode(bytes);
  return String.fromCharCode(...bytes);
}

const B64 = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

function base64Of(bytes: Uint8Array): string {
  let out = "";
  for (let i = 0; i < bytes.length; i += 3) {
    const a = bytes[i];
    const b = i + 1 < bytes.length ? bytes[i + 1] : 0;
    const c = i + 2 < bytes.length ? bytes[i + 2] : 0;
    const n = (a << 16) | (b << 8) | c;
    out += B64[(n >> 18) & 63] + B64[(n >> 12) & 63];
    out += i + 1 < bytes.length ? B64[(n >> 6) & 63] : "=";
    out += i + 2 < bytes.length ? B64[n & 63] : "=";
  }
  return out;
}

function base64ToBytes(b64: string): Uint8Array {
  const clean = b64.replace(/[^A-Za-z0-9+/=]/g, "");
  const out: number[] = [];
  for (let i = 0; i < clean.length; i += 4) {
    const vals = [...clean.slice(i, i + 4)].map((ch) => (ch === "=" ? 0 : B64.indexOf(ch)));
    const n = (vals[0] << 18) | (vals[1] << 12) | (vals[2] << 6) | vals[3];
    out.push((n >> 16) & 255);
    if (clean[i + 2] !== "=") out.push((n >> 8) & 255);
    if (clean[i + 3] !== "=") out.push(n & 255);
  }
  return Uint8Array.from(out);
}
